fix 礼包无服务器兑换英文文档 (#861)
taptap/tds-docs: i18n/en/docusaurus-plugin-content-docs/current/sdk/tds-gift/no-server-guide.mdx

---
title: Serverless Exchange Development Guide
sidebar_label: Serverless Exchange
sidebar_position: 3
---

:::tip

> This article is to explain the access process of the serverless redemption method, if you need the complete development content please see [**Guide**](/sdk/tds-gift/guide/).

Reviewer: ```suggestion
This article is to explain the access process of serverless redemption method. If you need complete development conten, please refer to [**Guide**](/sdk/tds-gift/guide/).
```

:::

## Present

> Docking gift redemption code requires the game party to maintain one or two web interfaces is relatively cumbersome, solid we simplify the redemption process as follows:

Reviewer: ```suggestion
Docking gift redemption code requires the game party to maintain one or two web interfaces, which is relatively cumbersome. Therefore, we have simplified the redemption process as follows:
```

> ** The game party requests the redemption interface for redemption and judges whether the redemption is successful based on the information returned by the interface. **

Reviewer: ```suggestion
** The game party requests the redemption interface for redemption, and judges whether the redemption is successful based on the information returned by the interface. **
```

> This way, the player only needs to call one interface to complete the redemption process.

Reviewer: ```suggestion
In this way, the game party only needs to call one interface to complete the redemption process.
```

:::tip

> The client receives the redemption result directly, so the security of this method is not as high as other methods.

Reviewer: ```suggestion
The exchange results received directly by the client can be intercepted, so the security of this method is not as high as other methods.
```

:::

## Matchmaking

### Redemption interface

The domain name is divided into two addresses according to the developer centre, and the gift pack data of the two addresses are not interoperable:

| TapTap Version     | Domain Name                      |
| --------------- | ------------------------- |
| China taptap.cn | https://poster-api.xd.cn  |

> | Other taptap.io  | https://poster-api.xd.com |

Reviewer: ```suggestion
| Overseas taptap.io  | https://poster-api.xd.com |
```

The redemption interface parameters are described in detail in the [**Developer's Guide - Serverless Redemption**](/sdk/tds-gift/guide/#redeeming-without-a-server-1).

> A signature must be generated when initiating a redemption. The signature of the service-less redemption interface is calculated from the clientId and other information, see [**Documentation**](/sdk/tds-gift/guide/#signature) for the signature calculation process.

Reviewer: ```suggestion
A signature need to be generated when initiating a redemption. The signature of the service-less redemption interface is calculated from the Client Id and other information, see [**Documentation**](/sdk/tds-gift/guide/#signature) for the signature calculation process.
```

### Return value

#### Result

Judge whether the exchange behaviour is successful or not by whether the error field in the return value is 0. If the error field is judged not to be 0, you can see the specific reason according to [**error-codes**](/sdk/tds-gift/guide/#error-codes).

#### Verification

> To prevent the return information from being tampered with, it is recommended that you do the following

Reviewer: ```suggestion
To prevent the return information from being tampered with, it is recommended that you take the following measures:
```

> 1. To prevent replay attacks, check that the return timestamp is not too different from the client's time;

Reviewer: ```suggestion
1. To prevent replay attacks, you need to verify whether the return timestamp is too different from the client's time;
```

2. Calculate a signature based on the return value and verify that it matches the signature in the return value. The signature calculation process is described in [**Documentation**](/sdk/tds-gift/guide/#signature);
2. It is recommended that the name of the gift content can be an encrypted name.

#### Prize content

> According to the content or content_obj field in the return value, we can get the content of the prize corresponding to the redemption code, but the difference is that content is a nested json string and parsing it directly as an object may return an error. 

Reviewer: ```suggestion
The prize content corresponding to the redemption code can be obtained based on the content or content_obj field in the return value. The difference is that content is a nested json string and parsing it directly as an object may result in an error. 
```

> The content of the prize is not limited to English or Chinese, so it's best to maintain a table for the prize to make it easier for the developer to judge.

Reviewer: ```suggestion
The prize content setting is not limited to English and Chinese. In order to facilitate development and judgment, it is recommended to maintain a prop list.
```

## Other
### Development Environment

> To use the serverless redemption interface to differentiate the official test environment, you will need to create a new application in the Developer Centre to complete it.

Reviewer: ```suggestion
Using the serverless redemption interface to distinguish the formal testing environment requires creating different applications in the Developer Center.
```

> If it is not possible to deal with environmental issues by application, it is recommended to use the [without secondary check redemption interface](/sdk/tds-gift/guide/#without-secondary-check) by configuring the server information, configure the server code forwarded to different environments to complete the redemption.

Reviewer: ```suggestion
If it is not possible to deal with environmental issues by application, it is recommended to use the [without secondary check redemption interface](/sdk/tds-gift/guide/#without-secondary-check) to complete the redemption by configuring server information and configuring the server code to be forwarded to different environments .
```
